components/sup/src/cli.rs

```diff
@@ -0,0 +1,223 @@
+use clap::App;
+
+use hab::cli::{
+    sub_sup_bash, sub_sup_depart, sub_sup_run, sub_sup_secret, sub_sup_sh, sub_sup_term,
+    sub_svc_status,
+};
+use VERSION;
+
+pub fn cli<'a, 'b>() -> App<'a, 'b> {
+    clap_app!(("hab-sup") =>
+        (about: "The Habitat Supervisor")
+        (version: VERSION)
+        (author: "\nAuthors: The Habitat Maintainers <humans@habitat.sh>\n")
+        // set custom usage string, otherwise the binary
+        // is displayed confusingly as `hab-sup`
+        // see: https://github.com/kbknapp/clap-rs/blob/2724ec5399c500b12a1a24d356f4090f4816f5e2/src/app/mod.rs#L373-L394
+        (usage: "hab sup <SUBCOMMAND>")
+        (@setting VersionlessSubcommands)
+        (@setting SubcommandRequiredElseHelp)
+        // this is the _full_ list of supervisor related cmds
+        // they are all enumerated here so that the entire help menu
+        // can be displayed from `hab sup --help`
+        (subcommand: sub_sup_bash().aliases(&["b", "ba", "bas"]))
+        (subcommand: sub_sup_depart().aliases(&["d", "de", "dep", "depa", "depart"]))
+        (subcommand: sub_sup_run().aliases(&["r", "ru"]))
+        (subcommand: sub_sup_secret().aliases(&["sec", "secr"]))
+        (subcommand: sub_sup_sh().aliases(&[]))
+        (subcommand: sub_svc_status().aliases(&["stat", "statu"]))
+        (subcommand: sub_sup_term().aliases(&["ter"]))
+    )
+}
+
+#[cfg(test)]
+mod test {
+    use super::cli;
+
+    macro_rules! assert_cli_cmd {
+        ($test:ident, $cmd:expr, $( $key:expr => $value:tt ),+) => {
+            #[test]
+            fn $test() {
+                assert_cmd!(cli(), $cmd, $( $key => $value ),+ );
+            }
+        }
+    }
+
+    mod sup_run {
+        use super::*;
+
+        assert_cli_cmd!(should_take_listen_gossip_flag,
```

Comment:
I think tests like this are a bit overkill. Considering the code that establishes this behavior is only
```rust
        (@arg LISTEN_GOSSIP: --("listen-gossip") +takes_value {valid_socket_addr}
            "The listen address for the Gossip System Gateway. If not specified, the value will \
            be taken from the HAB_LISTEN_GOSSIP environment variable if defined. [default: 0.0.0.0:9638]")
```
I don't know that it calls for having a separate test that only exercises the CLI parsing, not the behavior that the argument should result in.

I could see some tests to exercise that `valid_socket_addr` behaves as expected (and that would have more value since several options depend upon it), but I think all the `should_take…` tests are adding code without sufficient value.

Reply:
Generally I would agree with you. But the thinking is that we will be moving away from the `clap_app!` macro for defining the cli as we bring everything into a single binary. Without tests for cli parsing it would be tough to know if we are getting the same cli behavior as before.